### PR TITLE
Testing architecture

### PR DESCRIPTION
**Pull request by LesiaDorofiienko**

### CODE REVIEW

File: src/client/api-client.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { AuthController, CarController, UserController } from "../controllers";
+import { CookieJar } from "tough-cookie";
+import { config } from "../../config/config.js";
+
+export class APIClient {
+  readonly auth: AuthController;
+  readonly cars: CarController;
+  readonly users?: UserController;
```
Comment (sttaran) [inline src/client/api-client.ts]: you either have it or not

Comment (sttaran) [reply src/client/api-client.ts]: if not then remove declaration

File: src/controllers/base.controller.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { CookieJar } from "tough-cookie";
+import { wrapper } from "axios-cookiejar-support";
+import axios, { AxiosInstance } from "axios";
+import { config } from "../../config/config";
+
+export class BaseController {
+  readonly client: AxiosInstance;
+
+  constructor(
```
Comment (sttaran) [inline src/controllers/base.controller.ts]: better use "object of parameters"

File: src/fixtures/index.ts [hunk 3]
```diff
@@ -16,6 +20,9 @@ export type Fixtures = {
   userGaragePage: GaragePage;
   pageWithAuth: Page;
   userAPIClient: APIRequestContext;
+  client: APIClient;
+  clientWithUser: (userData: any) => Promise<APIClient>;
+  clientWithNewUser: APIClient;
```
Comment (sttaran) [inline src/fixtures/index.ts:25]: userData can have type, you know it's type

File: tests/api/brand-fixtures.spec.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import { test } from "../../src/fixtures/";
+import { expect } from "@playwright/test";
+import { VALID_BRANDS_RESPONSE_BODY } from "../../src/data/dict/brands";
+import { VALID_BRAND_MODELS } from "../../src/data/dict/models";
+import { USERS } from "../../src/data/dict/users";
+
+test.describe.skip("API", () => {
+  test("should return user's cars", async ({ client }) => {
+    const response = await client.cars.getUserCars();
+    expect(response.status, "Status code should be 200").toEqual(200);
```
Comment (sttaran) [inline tests/api/brand-fixtures.spec.ts:10]: check response body

Comment (sttaran) [reply tests/api/brand-fixtures.spec.ts:10]: you should know what exactly it returns

File: tests/api/cars-controller.spec.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    });
+
+    test("should return user cars", async ({ clientWithNewUser }) => {
+      await clientWithNewUser.cars.createCar(validCarData);
+
+      const response = await clientWithNewUser.cars.getUserCars();
+
+      expect(response.status).toEqual(200);
+      expect(response.data.status).toEqual("ok");
+      expect(response.data.data[0]).toMatchObject(validCarData);
```
Comment (sttaran) [inline tests/api/cars-controller.spec.ts:104]: check all cars in response, otherwise you do not know what it returns

**Review by sttaran — CHANGES_REQUESTED** [review]
Group tests with files. It's hard to read all tests in one file

File: tests/api/brand-fixtures.spec.ts (hunk 4 above)
Comment (LesiaDorofiienko) [reply tests/api/brand-fixtures.spec.ts:10]: This is code from the lesson

File: src/fixtures/index.ts (hunk 3 above)
Comment (sttaran) [reply src/fixtures/index.ts:25]: probably you've forgotten this one

Comment (sttaran) [reply src/fixtures/index.ts:25]: I meant do no use any
`(userData: any)`

File: tests/api/brands-controller.spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      email: USERS.lesia.email,
+      password: USERS.lesia.password,
+      remember: false,
+    });
+  });
+
+  test("should return user's cars", async () => {
+    const response = await client.cars.getUserCars();
+    expect(response.status, "Status code should be 200").toEqual(200);
+    // expect(response.data, "Valid brands should be returned").toEqual(VALID_BRANDS_RESPONSE_BODY)
```
Comment (sttaran) [inline tests/api/brands-controller.spec.ts:23]: why don't you check response body?

File: tests/api/cars-controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      expect(response.data).toEqual(VALID_MODELS_RESPONSE);
+    });
+
+    test("should return model by id", async ({ client }) => {
+      const modelId = VALID_MODELS_RESPONSE.data[0].id;
+      const response = await client.cars.getModelsById(modelId);
+
+      expect(response.status).toEqual(200);
+      expect(response.data.data).toEqual(VALID_MODELS_RESPONSE.data[0]);
+    });
```
Comment (sttaran) [inline tests/api/cars-controller.spec.ts:51]: check all models by id

**Review by sttaran — COMMENTED** [review]
1. Prepare state for tests in before* hooks and clear data in after* hooks
2. It is better to goup tests with test files. (createCar.specs.js, updateCar.specs.js etc)

**Pull request merged**